### PR TITLE
Update colour coding of DisputeStatusChip based on dispute status and urgency

### LINKED ISSUE #6722: Dispute list: colour code `Status` column pills to highlight urgency
Part of #6355 

This Status Chip that appears in the Disputes List screen and various other screens should be updated as per the following criteria:

* red = needs response within 72h
* yellow = needs response > 72h
* green = won
* blue = in progress
* grey = lost

### PR DESCRIPTION
**Pull request by Jinksi**
Fixes #6722 

#### Changes proposed in this Pull Request

This PR changes the colour of the Dispute Status Chip based on the following criteria:

* red = needs response within 72h
* yellow = needs response > 72h
* green = won
* blue = in progress / under review
* grey = lost

This Dispute Status Chip is rendered in the following locations, each with updated colours:

**Disputes List**
image

**Dispute Details**
image

**Transaction Details**
image

#### Testing instructions

* Ensure your store has multiple disputes in various states, e.g. Lost, Won, Needs Response.
	* Possible statuses: `warning_needs_response`, `warning_under_review`, `warning_closed`, `needs_response`, `under_review`, `charge_refunded`, `won`, `lost`
	* Submit dispute evidence with `winning_evidence` in the `Additional Details` field to create a `Won` dispute. [See wiki](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows#winning-dispute).
* Navigate to Payments → Disputes. Ensure the colour of the chip under the **Status** column match the criteria listed above.
* Click on a dispute to view the Dispute Details screen. Ensure the status chip colour matches the criteria listed above.
* Click on the charge to view the Transaction Details screen. Ensure the status chip colour matches the criteria listed above.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6723` or branch name `fix/6722-dispute-list-status-chip-colours` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/6722-dispute-list-status-chip-colours) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 3ba247c29d38f7704e3817ef06720bef4593e1ac
- Build time: 2023-07-13 02:03:04 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/components/chip/test/index.test.tsx
Comment (Jinksi) [inline client/components/chip/test/index.test.tsx:1]: Refactored this test file to TypeScript. No changes to logic expected.

File: client/components/dispute-status-chip/test/index.test.tsx
Comment (Jinksi) [inline client/components/dispute-status-chip/test/index.test.tsx:1]: Refactored this test file to TypeScript. No changes to logic expected.

File: client/components/chip/style.scss
Comment (Jinksi) [inline client/components/chip/style.scss:1]: Note: all of these background colours (e.g. `$studio-blue-5`) are darker than the Disputes Flows designs in Figma (which uses e.g. `$studio-blue-0` ).

I will confirm with design if we want to change these WCPay-wide and open a new issue if relevant.

image

Comment (Jinksi) [reply client/components/chip/style.scss:1]: Asked in Slack: p1689057662945359-slack-C03LA109BKM

File: client/disputes/utils.ts
Comment (Jinksi) [inline client/disputes/utils.ts:27]: Ideally, use this shared util function throughout disputes codebase. Will handle this in followup PR.

File: client/components/chip/style.scss
Comment (Jinksi) [reply client/components/chip/style.scss:1]: Will address in a new PR.

File: client/components/dispute-status-chip/mappings.ts
Comment (Jinksi) [inline client/components/dispute-status-chip/mappings.ts:1]: This file contains the core of the changes, aside from the urgent disputes needing a response = red.

Needs response = 🟡 Yellow
`needs_response`: warning
`warning_needs_response`: warning

In progress = 🔵 Blue
`warning_under_review`: primary
`under_review`: primary

Lost = ⚪️ Grey
`warning_closed`: light
`charge_refunded`: light
`lost`: light

Won = 🟢 Green
`won`: success

File: client/components/chip/style.scss
Comment (Jinksi) [inline client/components/chip/style.scss:15]: Added new chip type/colour: success = green.

File: changelog/fix-6722-dispute-list-status-chip-colours
Comment (shendy-a8c) [inline changelog/fix-6722-dispute-list-status-chip-colours:4]: In https://github.com/Automattic/woocommerce-payments/issues/6355, there is a task "Add Changelog entry for this epics changes", so I think for this PR, no need to add changelog entry as it'd be done in a separate PR.

**Review by shendy-a8c — COMMENTED** [review]
There is a [local function named isDueWithin()](https://github.com/Automattic/woocommerce-payments/blob/fix/6722-dispute-list-status-chip-colours/client/overview/task-list/dispute-tasks.ts#L28). Do you think that needs to be removed and replaced with this new `isDueWithin()` from `client/disputes/utils.ts`?

File: client/disputes/utils.ts
Comment (shendy-a8c) [inline client/disputes/utils.ts]: I don't see any `isDueWithin()` call that does not pass `dueBy`.

```suggestion
	dueBy: CachedDispute[ 'due_by' ] | EvidenceDetails[ 'due_by' ];
```

File: changelog/fix-6722-dispute-list-status-chip-colours
Comment (Jinksi) [reply changelog/fix-6722-dispute-list-status-chip-colours:4]: Ah, that's right. It's so automatic to add a changelog 😆

File: client/disputes/utils.ts
Comment (shendy-a8c) [inline client/disputes/utils.ts]: Got to ask again as per my understanding, 'within' is inclusive, so when diff is equal to the value of `days`, it should be true.

```suggestion
	const isWithinDays = dueByMoment.diff( now, 'days', true ) <= days;
```

**Comment by Jinksi** [comment]
> There is a [local function named isDueWithin()](https://github.com/Automattic/woocommerce-payments/blob/fix/6722-dispute-list-status-chip-colours/client/overview/task-list/dispute-tasks.ts?rgh-link-date=2023-07-13T01%3A19%3A28Z#L28). Do you think that needs to be removed and replaced with this new isDueWithin() from client/disputes/utils.ts?

I initially has moved this function to `client/disputes/utils.ts` but I reverted this to keep this PR streamlined, especially due to the 6.2 deadline.

image

I was planning to clean up duplicated logic calculating due_by times in a PR next week. What do you think?

**Review by shendy-a8c — APPROVED** [review]
Looks good and tested well! I left some feedbacks but none is crucial that should block the merge.

**Comment by shendy-a8c** [comment]
> I was planning to clean up duplicated logic calculating due_by times in a PR next week. What do you think?

Sounds good.

File: client/disputes/utils.ts
Comment (Jinksi) [reply client/disputes/utils.ts]: Thanks!

File: client/disputes/utils.ts
Comment (Jinksi) [reply client/disputes/utils.ts]: I agree – let's go with within is `<=`.

Updated this logic to your suggestion `<=` and made the unit tests explicit about this in 3ba247c.

File: changelog/fix-6722-dispute-list-status-chip-colours
Comment (Jinksi) [reply changelog/fix-6722-dispute-list-status-chip-colours:4]: 🤦 I didn't change this before merge. I've added this to the changelog task item on the epic #6355

**Comment by htdat** [comment]
👋 @Jinksi and @shendy-a8c - it looks like this PR is causing this E2E test suite failure: 


> FAIL tests/e2e/specs/wcpay/merchant/merchant-disputes-submit-winning.spec.js (123.065 s)
> Disputes > Submit winning dispute
> ✕ should process and confirm a winning dispute (100010 ms)
> 
> ● Disputes > Submit winning dispute › should process and confirm a winning dispute

It appears in the last commit https://github.com/Automattic/woocommerce-payments/commit/3ba247c29d38f7704e3817ef06720bef4593e1ac here as well as since this PR was merged to `develop` in https://github.com/Automattic/woocommerce-payments/commits/develop

Markup on 2023-07-14 at 10:24:01

**Comment by Jinksi** [comment]
Thanks @htdat - apologies for missing this. I hadn’t looked closely enough at the e2e tests since there had been unrelated tests failing all week 🤦‍♂️

I’ll get a fix ready.

**Comment by Jinksi** [comment]
PR #6787 ready for review

**Pull request merged**
